addons/knowledge/models/knowledge_article.py

```diff
@@ -0,0 +1,1423 @@
+        if vals_by_parent_id:
+            parent_ids = list(vals_by_parent_id.keys())
+            try:
+                self.check_access_rights('write')
+                self.env['knowledge.article'].browse(parent_ids).check_access_rule('write')
+            except AccessError:
+                raise AccessError(_("You cannot create an article under articles on which you cannot write"))
+            max_sequence_by_parent = self._get_max_sequence_inside_parents(parent_ids)
+
+        # update sequences
+        for parent_id, article_vals in vals_by_parent_id.items():
+            current_sequence = 0
+            if parent_id in max_sequence_by_parent:
+                current_sequence = max_sequence_by_parent[parent_id] + 1
+
+            for vals in article_vals:
+                if 'sequence' in vals:
+                    current_sequence = vals.get('sequence')
+                else:
+                    vals['sequence'] = current_sequence
+                    current_sequence += 1
+
+        # sort by sudo / not sudo
+        notsudo_articles = iter(super(Article, self).create([
+            vals for vals, can_sudo in zip(vals_list, vals_as_sudo)
+            if not can_sudo
+        ]))
+        sudo_articles = iter(super(Article, self.sudo()).create([
+            vals for vals, can_sudo in zip(vals_list, vals_as_sudo)
+            if can_sudo
+        ]).with_env(self.env))
+        articles = self.env['knowledge.article']
+        for vals, is_sudo in zip(vals_list, vals_as_sudo):
+            if is_sudo:
+                articles += next(sudo_articles)
+            else:
+                articles += next(notsudo_articles)
+
+        return articles
+
+    def write(self, vals):
+        # Move under a parent is considered as a write on it (permissions, ...)
+        _resequence = False
+        if vals.get('parent_id'):
+            parent = self.browse(vals['parent_id'])
+            try:
+                parent.check_access_rights('write')
+                parent.check_access_rule('write')
+            except AccessError:
+                raise AccessError(_("You cannot move an article under %(parent_name)s as you cannot write on it",
+                                    parent_name=parent.display_name))
+            if 'sequence' not in vals:
+                max_sequence = self._get_max_sequence_inside_parents(parent.ids).get(parent.id, -1)
+                vals['sequence'] = max_sequence + 1
+            else:
+                _resequence = True
+
+        result = super(Article, self).write(vals)
+
+        # resequence only if a sequence was not already computed based on current
+        # parent maximum to avoid unnecessary recomputation of sequences
+        if _resequence:
+            self._resequence()
+
+        return result
+
+    @api.returns('self', lambda value: value.id)
+    def copy(self, default=None):
+        default = dict(default or {},
+                       name=_("%s (copy)", self.name))
```

Comment:
calling `self.copy({'name':'New Name'})` is ignored then

Reply:
Trou

Reply:
c'est toi le trou

Reply:
That's wild.